### ISSUE TITLE
Shared memory not working with spawned process

### ISSUE THREAD

**Issue by rodarima** [open-mpi/ompi#6823]
After launching a new process with `MPI_Comm_spawn`, the call to allocate a shared memory region fails:

```
[mio:03625] *** An error occurred in MPI_Win_allocate_shared
[mio:03625] *** reported by process [3230072833,0]
[mio:03625] *** on communicator 
[mio:03625] *** MPI_ERR_RMA_SHARED: Memory cannot be shared
[mio:03625] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[mio:03625] ***    and potentially your MPI job)
[mio:03620] 1 more process has sent help message help-mpi-errors.txt / mpi_errors_are_fatal
[mio:03620] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

The following reproducer works with Intel MPI, buf fails with OpenMPI.

Parent code:
```c
#define _GNU_SOURCE
#include <sched.h>

#include <stdio.h>
#include <mpi.h>
#include <stdlib.h>
#include <unistd.h>
#include <assert.h>

int
main(int argc, char *argv[])
{
	int provided, n, rank, size;
	MPI_Comm intercomm, universe;
	MPI_Win win;
	int *buf;
	MPI_Aint bufsize;
	int disp = sizeof(int);
	int k;

	MPI_Init_thread(NULL, NULL, MPI_THREAD_MULTIPLE, &provided);
	assert(provided == MPI_THREAD_MULTIPLE);

	MPI_Comm_size(MPI_COMM_WORLD, &size);

	MPI_Comm_spawn("./worker", MPI_ARGV_NULL, 1,
		MPI_INFO_NULL, 0, MPI_COMM_WORLD,
		&intercomm, MPI_ERRCODES_IGNORE);

	MPI_Intercomm_merge(intercomm, 0, &universe);
	MPI_Comm_size(universe, &k);
	assert(k == 2);

	bufsize = sizeof(int);

	MPI_Win_allocate_shared(bufsize, 1, MPI_INFO_NULL, universe, &buf, &win);

	buf[0] = 666;

	MPI_Barrier(universe);

	/* Worker runs now */

	MPI_Barrier(universe);

	assert(buf[0] == 555);
	MPI_Finalize();

	return 0;
}
```

Worker code:
```c
#include <mpi.h>
#include <stddef.h>
#include <assert.h>


int
main(int argc, char *argv[])
{
	MPI_Comm parent, universe;
	int *buf;
	size_t bufsize;
	int rank, disp;
	char hostname[100];
	MPI_Win win;
	MPI_Aint asize;

	MPI_Init(&argc, &argv);

	MPI_Comm_get_parent(&parent);

	assert(parent != MPI_COMM_NULL);

	MPI_Intercomm_merge(parent, 0, &universe);


	MPI_Win_allocate_shared(0, 1, MPI_INFO_NULL, universe, &buf, &win);
	MPI_Win_shared_query(win, MPI_PROC_NULL, &asize, &disp, &buf);

	MPI_Barrier(universe);

	assert(buf[0] == 666);

	buf[0] = 555;

	MPI_Barrier(universe);

	MPI_Finalize();
	return 0;
}

```

Makefile:
```makefile
CC=mpicc

CFLAGS=-g -O0

BIN=main worker

all: $(BIN)

clean:
	rm -f $(BIN) *.o
```

To run, I use `mpirun --oversubscribe -n 1 ./main`.

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)

Release v4.0.1

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Arch Linux package install, build date Fri 29 Mar 2019.

### Please describe the system on which you are running

* Operating system/version: 5.1.16-arch1-1-ARCH

**Comment by hjelmn** [comment]
Are you sure all the processes are running locally?

**Comment by rodarima** [comment]
Yes, I'm using my laptop. Sorry I forgot to mention.


---- Nathan Hjelm escribió ----

**Comment by hjelmn** [comment]
If all the processes are local then the bug is in the internal ```ompi_group_have_remote_peers()``` function. I should be able to diagnose and fix it some time today.

**Comment by rodarima** [comment]
Thanks!

Also `MPI_Comm_split_type` fails to detect they run in the same host, using the shared memory flag.

---- Nathan Hjelm escribió ----

**Comment by hjelmn** [comment]
Yup. Figured as much. That uses the same underlying code.

**Comment by hjelmn** [comment]
I found the issue on the ompi side but the obvious fix doesn't work because PMIx doesn't provide the locality info right now. @rhc54 should it be providing locality for all local procs (even dpm). If not I can work around that.

**Comment by rhc54** [comment]
@hjelmn Just to be sure I understand: you would like PMIx to report locality for all procs on the node, even those in another "job" (since something started by comm_spawn is in its own job). Correct?

If so, that's no problem - could have it for you by next week if that's okay.

**Comment by hjelmn** [comment]
@rhc54 Exactly. That plus a change to ```ompi_proc_allocate()``` to get the locality should be sufficient to fix this bug.

**Comment by hjelmn** [comment]
Next week is ok.

**Comment by hjelmn** [comment]
@rhc54 I can confirm this is now fixed on master. Can the change be back-ported to the release braches?

**Comment by rhc54** [comment]
@hjelmn I think it should be okay for the OMPI v4 branch, but I don't see an easy way to go further back than that. Acceptable?

**Comment by goduck777** [comment]
I found this error seems to show up again in recent version.

I got the same error message as @rodarima a few month ago. After I downloaded the daily snapshot of openmpi v4.0.x source package and compiled it, the problem was solved then.

However, recently I tried to run my code on another cluster. I downloaded a more recent daily snapshot and compiled the code, but found the error reappeared. The error message looked exactly the same as @rodarima. So I went back to the previous cluster and did some tests.

The version I have tested,

openmpi-v4.0.x-202002100241-42acf4f work
openmpi-v4.0.x-202003240241-34c4f93 error
openmpi-v4.0.x-202005050241-3b600b8 error

So it seems some changes between 20200210 and 20200324 brought back this error.
